### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Pirakas Anthon

### DISCUSSION THREAD

**Discussion by apirakas** [heig-vd-dai-course/heig-vd-dai-course#83] [Show and tell]
**Repo:** https://github.com/apirakas/heig-vd-dai-course-java-ios-practical-content

**Generate different files with different sizes (1B, 1KiB, 1MiB, 5MiB). Compare the results with the execution time of the different types of streams. Which one is the most efficient for each use case?**

I've coded a loop of 5 read/write iterations and then averaged them for each case. To get average execution values. Without carrying out any tests, I'd say that it's better to use functions with buffers, which minimize the number of I/O operations and is therefore faster. 

Here are the code results (make with chatGPT):

| Write               | Buffer | Moyenne (ms) | Moyenne (ns)   |
| ------------------- | ------ | ------------ | -------------- |
| BINARY_1B.bin       | Non    | 0.000682     | 682000 ns      |
| BINARY_1B.bin       | Buffer | 0.000696     | 696000 ns      |
| BINARY_1024B.bin    | Non    | 45.002       | 45002000 ns    |
| BINARY_1024B.bin    | Buffer | 0.00072      | 720000 ns      |
| BINARY_1048576B.bin | Non    | 14750.2      | 14750200000 ns |
| BINARY_1048576B.bin | Buffer | 1.8          | 1800000 ns     |
| BINARY_5242880B.bin | Non    | 59091.2      | 59091200000 ns |
| BINARY_5242880B.bin | Buffer | 2.6          | 2600000 ns     |
| TEXT_1B.txt         | Non    | 0.000625     | 625000 ns      |
| TEXT_1B.txt         | Buffer | 0.002        | 2000 ns        |
| TEXT_1024B.txt      | Non    | 0.002        | 2000 ns        |
| TEXT_1024B.txt      | Buffer | 0.002        | 2000 ns        |
| TEXT_1048576B.txt   | Non    | 39.4         | 39400000 ns    |
| TEXT_1048576B.txt   | Buffer | 27.4         | 27400000 ns    |
| TEXT_5242880B.txt   | Non    | 100.2        | 100200000 ns   |
| TEXT_5242880B.txt   | Buffer | 110.2        | 110200000 ns   |

| Read                | Buffer | Moyenne (ms) | Moyenne (ns) |
| ------------------- | ------ | ------------ | ------------ |
| BINARY_1B.bin       | Non    | 0.000618     | 618000 ns    |
| BINARY_1B.bin       | Buffer | 0.000053     | 53000 ns     |
| BINARY_1024B.bin    | Non    | 0.000625     | 625000 ns    |
| BINARY_1024B.bin    | Buffer | 0.000042     | 42000 ns     |
| BINARY_1048576B.bin | Non    | 0.000617     | 617000 ns    |
| BINARY_1048576B.bin | Buffer | 0.000041     | 41000 ns     |
| BINARY_5242880B.bin | Non    | 0.000607     | 607000 ns    |
| BINARY_5242880B.bin | Buffer | 0.000042     | 42000 ns     |
| TEXT_1B.txt         | Non    | 0.000623     | 623000 ns    |
| TEXT_1B.txt         | Buffer | 0.000058     | 58000 ns     |
| TEXT_1024B.txt      | Non    | 0.0014       | 1400000 ns   |
| TEXT_1024B.txt      | Buffer | 0.000898     | 898000 ns    |
| TEXT_1048576B.txt   | Non    | 75.4         | 75400000 ns  |
| TEXT_1048576B.txt   | Buffer | 11.4         | 11400000 ns  |
| TEXT_5242880B.txt   | Non    | 255.4        | 255400000 ns |
| TEXT_5242880B.txt   | Buffer | 14.2         | 14200000 ns  |

**Which type of stream is the most efficient for each use case?**

In general, accessing buffers is more efficient, but for writing text the 2 methods have similar performance, but I think I've made a mistake, but I can't see where.

**Why is it more efficient than the other types of streams?**

as I said before, thanks to I/O access optimizations with buffers.

**What is the difference between binary data and text data?**

Binary data consists of 0s and 1s, while text data consists of readable characters.

**What is a character encoding?**

A character encoding is a method that assigns numeric values to characters for computer storage and processing.

**Why is this methodology important?**

This methodology is important because it provides a structured and  systematic approach to solving problems or conducting research, ensuring consistency, reliability, and the ability to replicate results.

**Comment by ludelafo** [comment]
Great! Thank you.

> but for writing text the 2 methods have similar performance, but I think I've made a mistake, but I can't see where.

From what I saw in your repository, everything seems right. My guess is that for such small files, it would be difficult to see the changes but for larger files, the buffered version should be way better.

> This methodology is important because it provides a structured and systematic approach to solving problems or conducting research, ensuring consistency, reliability, and the ability to replicate results.

What a simple and concise explanation. Great job!
